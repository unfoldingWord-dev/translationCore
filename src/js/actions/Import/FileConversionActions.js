/**
 * @Description:
 * Actions that call helpers to handle business logic for conversions
**/

export const convert = () => {
  // return((dispatch, getState) => {
  //   FileConversionHelpers.convertToProject(param);
  // });
};
