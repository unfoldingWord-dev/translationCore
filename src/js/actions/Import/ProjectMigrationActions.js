/**
 * @Description:
 * Actions that call helpers to handle business logic for migtaions
**/

export const migrate = () => {
  return((dispatch, getState) => {
  });
};
