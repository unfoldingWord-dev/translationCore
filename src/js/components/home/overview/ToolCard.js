import React, { Component } from 'react'
import TemplateCard from './TemplateCard'

class ToolCard extends Component {

  heading(callback) {
    let link = this.content() ? <a onClick={callback}>Change Tool</a> : <a></a>
    return (
      <span>Current Tool {link}</span>
    );
  }

  content() {
    let content;
    let { toolName } = this.props.reducers.toolsReducer;
    if (toolName) {
      content = (
        <div>
          <strong>{toolName}</strong>
          <p>details</p>
        </div>
      );
    }
    return content;
  }

  disabled() {
    let { projectSaveLocation } = this.props.reducers.projectDetailsReducer;
    return !projectSaveLocation;
  }

  render() {
    let emptyMessage = 'Select a tool';
    let emptyButtonLabel = 'Tool';
    let emptyButtonOnClick = () => { this.props.actions.goToNextStep() };
    return (
      <TemplateCard
        heading={this.heading(emptyButtonOnClick)}
        content={this.content()}
        emptyMessage={emptyMessage}
        emptyButtonLabel={emptyButtonLabel}
        emptyButtonOnClick={emptyButtonOnClick}
        disabled={this.disabled()}
      />
    )
  }
}

export default ToolCard
