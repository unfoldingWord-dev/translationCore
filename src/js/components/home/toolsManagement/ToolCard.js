import React, { Component } from 'react';
import PropTypes from 'prop-types';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import { Card, CardHeader } from 'material-ui';
import { Glyphicon } from 'react-bootstrap';
import { connect } from 'react-redux';
import _ from 'lodash';
// helpers
import * as ToolCardHelpers from '../../../helpers/ToolCardHelpers';
import { getTranslation } from '../../../helpers/localizationHelpers';
import { getGatewayLanguageList, hasValidOL } from '../../../helpers/gatewayLanguageHelpers';
// components
import Hint from '../../Hint';
import {
  getProjectBookId,
  getSetting,
  getSelectedToolName,
} from '../../../selectors';
import {
  WORD_ALIGNMENT, TRANSLATION_WORDS, TRANSLATION_NOTES,
} from '../../../common/constants';
import ToolCardBoxes from './ToolCardBoxes';
import ToolCardProgress from './ToolCardProgress';
import GlDropDownList from './GlDropDownList';
import ToolCardNotificationBadges from './ToolCardNotificationBadges';
// selectors
// consts

class ToolCard extends Component {
  constructor(props) {
    super(props);
    this.selectionChange = this.selectionChange.bind(this);
    this.handleSelect = this.handleSelect.bind(this);
    this.onLaunchClick = this.onLaunchClick.bind(this);
    this.loadProgress = _.debounce(this.loadProgress.bind(this), 200);
    this.state = {
      showDescription: false,
      progress: 0,
      selectedCategoriesChanged: false,
      glSelectedChanged: false,
      selectedGL: '',
      gatewayLanguageList: null,
    };
  }

  componentWillMount() {
    let {
      tool,
      bookId,
      glSelected: selectedGL,
    } = this.props;
    const gatewayLanguageList = getGatewayLanguageList(bookId, tool.name);

    // if there is only one gateway Language then select it as the GL for the tool card.
    if (gatewayLanguageList.length === 1) {
      selectedGL = gatewayLanguageList[0].code;
    }
    this.selectionChange(selectedGL);
    this.setState({ selectedGL, gatewayLanguageList });
  }

  componentDidMount() {
    this.loadProgress();
  }

  componentDidUpdate(prevProps) {
    const glSelectedChanged = prevProps.glSelected !== this.props.glSelected;

    if (glSelectedChanged) {
      this.setState({ glSelectedChanged });
    }

    if (!_.isEqual(prevProps.selectedCategories, this.props.selectedCategories)) {
      this.setState({ selectedCategoriesChanged: true });
      this.loadProgress();
    }
  }

  loadProgress() {
    const { tool, actions } = this.props;
    const { progress } = this.state;

    setTimeout(() => {
      const results = {};
      // TODO: This is an antipattern. Should update the state and use the prop coming from the state instead of assigning the result to an argument.
      actions.getProjectProgressForTools(tool.name, results);
      const toolProgress = results.progress;

      if (progress !== toolProgress) {
        this.setState({ progress: toolProgress ? toolProgress : 0 });
      }
    }, 0);
  }

  selectionChange(selectedGL) {
    if (selectedGL && selectedGL.trim()) {
      this.props.actions.setProjectToolGL(this.props.tool.name, selectedGL);
      this.setState({ selectedGL });
      this.loadProgress();
    }
  }

  getLaunchDisableMessage(id, developerMode, translate, name, selectedCategories) {
    const toolsWithCategories = [TRANSLATION_WORDS , TRANSLATION_NOTES];
    let launchDisableMessage = ToolCardHelpers.getToolCardLaunchStatus(this.state.selectedGL, id, developerMode, translate);

    if (!launchDisableMessage) { // if no errors, make sure we have original language
      const olBookPath = hasValidOL(id);

      if (!olBookPath) {
        launchDisableMessage = translate('tools.book_not_supported');
      }
    }

    if (!launchDisableMessage && !developerMode) { // if no errors and not developer mode , make sure we have a gateway language
      const { gatewayLanguageList } = this.state;
      launchDisableMessage = (gatewayLanguageList && gatewayLanguageList.length) ? null : translate('tools.book_not_supported');
    }

    if (!launchDisableMessage && (toolsWithCategories.includes(name) && selectedCategories.length === 0)) {
      launchDisableMessage = translate('tools.no_checks_selected');
    }

    return launchDisableMessage;
  }

  /**
   * Handles selecting the tool
   */
  handleSelect() {
    const { onSelect, tool } = this.props;
    onSelect(tool.name);
  }

  onLaunchClick() {
    const {
      isOLBookVersionMissing,
      toggleHomeView,
      selectedToolName,
      tool,
      actions: { warnOnInvalidations },
    } = this.props;
    const { selectedCategoriesChanged, glSelectedChanged } = this.state;
    const newSelectedToolName = tool.name;

    if (isOLBookVersionMissing) {
      // Show dialog with option to download missing resource
      this.props.onMissingResource();
    } else if (selectedToolName && !glSelectedChanged && !selectedCategoriesChanged && (selectedToolName === newSelectedToolName)) {
      // Show tool (Without loading tool data)
      toggleHomeView(false);
      warnOnInvalidations(newSelectedToolName);
    } else {
      // Load tool data then show tool
      this.handleSelect();
    }
  }

  render() {
    const {
      tool,
      bookId,
      translate,
      developerMode,
      actions: {
        updateCategorySelection,
        updateSubcategorySelection,
      },
      selectedCategories,
      availableCategories,
    } = this.props;
    const {
      progress, selectedGL, gatewayLanguageList,
    } = this.state;
    const launchDisableMessage = this.getLaunchDisableMessage(bookId, developerMode, translate, tool.name, selectedCategories);
    let desc_key = null;
    let showCheckBoxes = false;

    switch (tool.name) {
    case WORD_ALIGNMENT:
      desc_key = 'tools.alignment_description';
      break;

    case TRANSLATION_WORDS:
      showCheckBoxes = true;
      desc_key = 'tools.tw_part1_description';
      break;

    case TRANSLATION_NOTES:
      showCheckBoxes = true;
      break;

    default:
      break;
    }

    let descriptionLocalized = tool.description;

    if (desc_key) {
      descriptionLocalized = getTranslation(translate, desc_key, tool.description);
    }

    return (
      <MuiThemeProvider>
        <Card style={{ margin: '6px 0px 10px' }}>
          <img
            alt={tool.name}
            style={{
              float: 'left', height: '90px', margin: '10px',
            }}
            src={tool.badge}
          />
          <CardHeader
            title={tool.title}
            titleStyle={{ fontWeight: 'bold' }}
            subtitle={tool.version}
            style={{ display: 'flex', justifyContent: 'space-between' }}>
            <ToolCardNotificationBadges tool={tool} translate={translate} selectedCategories={selectedCategories} />
          </CardHeader><br />
          <ToolCardProgress progress={progress} />
          {
            showCheckBoxes &&
            <ToolCardBoxes
              key={selectedGL}
              toolName={tool.name}
              selectedCategories={selectedCategories}
              availableCategories={availableCategories}
              onCategoryChecked={updateCategorySelection}
              onSubcategoryChecked={updateSubcategorySelection}
              bookId={bookId}
              translate={translate}
              selectedGL={selectedGL}
            />
          }
          {this.state.showDescription ?
            (<div>
              <span style={{
                fontWeight: 'bold', fontSize: '16px', margin: '0px 10px 10px',
              }}>{translate('tools.description')}</span>
              <p style={{ padding: '10px' }}>
                {descriptionLocalized}
              </p>
            </div>) : (<div />)
          }
          <div style={{
            display: 'flex', justifyContent: 'space-between', flexDirection: 'row', alignItems: 'center',
          }}>
            <div style={{ display: 'flex', justifyContent: 'space-between' }}>
              <div
                style={{
                  padding: '10px 10px 0px', fontSize: '18px', cursor: 'pointer',
                }}
                onClick={() => this.setState({ showDescription: !this.state.showDescription })}
              >
                <span>{this.state.showDescription ? translate('tools.see_less') : translate('tools.see_more')}</span>
                <Glyphicon
                  style={{ fontSize: '18px', margin: '0px 0px 0px 6px' }}
                  glyph={this.state.showDescription ? 'chevron-up' : 'chevron-down'}
                />
              </div>
            </div>
            <GlDropDownList
              translate={translate}
              selectedGL={this.state.selectedGL}
              selectionChange={this.selectionChange}
              gatewayLanguageList={gatewayLanguageList}
            />
            <Hint
              position={'left'}
              size='medium'
              label={launchDisableMessage}
              enabled={launchDisableMessage ? true : false}
            >
              <button
                disabled={launchDisableMessage ? true : false}
                className='btn-prime'
                onClick={this.onLaunchClick}
                style={{ width: '90px', margin: '10px' }}
              >
                {translate('buttons.launch_button')}
              </button>
            </Hint>
          </div>
        </Card>
      </MuiThemeProvider>
    );
  }
}

ToolCard.propTypes = {
  onSelect: PropTypes.func.isRequired,
  translate: PropTypes.func.isRequired,
  bookId: PropTypes.string.isRequired,
  tool: PropTypes.object.isRequired,
  developerMode: PropTypes.bool.isRequired,
  actions: PropTypes.shape({
    getProjectProgressForTools: PropTypes.func.isRequired,
    setProjectToolGL: PropTypes.func.isRequired,
    updateSubcategorySelection: PropTypes.func.isRequired,
    updateCategorySelection: PropTypes.func.isRequired,
    warnOnInvalidations: PropTypes.func.isRequired,
  }),
  selectedCategories: PropTypes.array.isRequired,
  availableCategories: PropTypes.object.isRequired,
  isOLBookVersionMissing: PropTypes.bool.isRequired,
  onMissingResource: PropTypes.func.isRequired,
  selectedToolName: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.bool,
  ]),
  toggleHomeView: PropTypes.func.isRequired,
  glSelected: PropTypes.string.isRequired,
};

ToolCard.contextTypes = { store: PropTypes.any };

const mapStateToProps = (state) => ({
  bookId: getProjectBookId(state),
  developerMode: getSetting(state, 'developerMode'),
  selectedToolName: getSelectedToolName(state),
});

export default connect(mapStateToProps)(ToolCard);
