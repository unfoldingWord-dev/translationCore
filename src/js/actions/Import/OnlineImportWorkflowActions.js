import path from 'path-extra';
import fs from 'fs-extra';
// actions
import consts from '../../actions/ActionTypes';
import { deleteImportsFolder, deleteProjectFromImportsFolder } from '../../helpers/Import/ProjectImportFilesystemHelpers';
import * as AlertModalActions from '../../actions/AlertModalActions';
import * as OnlineModeConfirmActions from '../../actions/OnlineModeConfirmActions';
import * as ProjectImportStepperActions from '../ProjectImportStepperActions';
import * as MyProjectsActions from '../MyProjects/MyProjectsActions';
import * as ProjectDetailsActions from '../ProjectDetailsActions';
import * as ProjectInformationCheckActions from '../ProjectInformationCheckActions';
// helpers
import * as TargetLanguageHelpers from '../../helpers/TargetLanguageHelpers';
import {
  generateImportPath,
  verifyThisIsTCoreOrTStudioProject,
} from '../../helpers/Import/OnlineImportWorkflowHelpers';
import * as CopyrightCheckHelpers from '../../helpers/CopyrightCheckHelpers';
import {
  getProjectManifest,
  getProjectSaveLocation,
  getTranslate,
  getUsername,
} from '../../selectors';
import * as FileConversionHelpers from '../../helpers/FileConversionHelpers';
import * as ProjectDetailsHelpers from '../../helpers/ProjectDetailsHelpers';
import migrateProject from '../../helpers/ProjectMigration';
import Repo from '../../helpers/Repo';
import { isProjectSupported } from '../../helpers/ProjectValidation/ProjectStructureValidationHelpers';
import {
  openProject,
  closeProject,
  showInvalidVersionError,
} from '../MyProjects/ProjectLoadingActions';
import { delay } from '../../common/utils';
//constants
import { tc_MIN_VERSION_ERROR, IMPORTS_PATH } from '../../common/constants';
import * as ProjectImportFilesystemActions from './ProjectImportFilesystemActions';
import * as ProjectValidationActions from './ProjectValidationActions';

/**
 * Action that dispatches other actions to wrap up online importing
 */
export const onlineImport = () => (dispatch, getState) => new Promise((resolve, reject) => {
  const translate = getTranslate(getState());

  dispatch(OnlineModeConfirmActions.confirmOnlineAction(async () => {
    let importProjectPath = '';
    let link = '';

    try {
      await deleteImportsFolder();
      // Must allow online action before starting actions that access the internet
      link = getState().importOnlineReducer.importLink.trim();
      console.log('onlineImport() - link=' + link);

      dispatch(clearLink());
      // or at least we could pass in the locale key here.
      dispatch(AlertModalActions.openAlertDialog(translate('projects.importing_project_alert', { project_url: link }), true));

      const importPath = await generateImportPath(link);
      console.log('onlineImport() - import to: ' + importPath);

      await fs.ensureDir(importPath);
      console.log('onlineImport() - cloning repo into file system');
      await Repo.clone(link, importPath);
      const selectedProjectFilename = Repo.parseRemoteUrl(Repo.sanitizeRemoteUrl(link)).name;
      console.log('onlineImport() - selectedProjectFilename= ' + selectedProjectFilename);

      dispatch({ type: consts.UPDATE_SELECTED_PROJECT_FILENAME, selectedProjectFilename });
      importProjectPath = path.join(IMPORTS_PATH, selectedProjectFilename);

      // check if we can import the project
      console.log('onlineImport() - check if we can import the project');
      const isValid = verifyThisIsTCoreOrTStudioProject(importProjectPath);

      if (!isValid) {
        const errorMessage = translate('projects.online_import_error', { project_url: link, toPath: importProjectPath });
        console.warn('This is not a valid tStudio or tCore project we can migrate: ', errorMessage);
        throw errorMessage;
      }

      await isProjectSupported(importProjectPath, translate);
      const initialBibleDataFolderName = ProjectDetailsHelpers.getInitialBibleDataFolderName(selectedProjectFilename, importProjectPath);
      await migrateProject(importProjectPath, link, getUsername(getState()));
      // assign CC BY-SA license to projects imported from door43
      await CopyrightCheckHelpers.assignLicenseToOnlineImportedProject(importProjectPath);
      console.log('onlineImport() - start project validation');
      dispatch(ProjectValidationActions.initializeReducersForProjectImportValidation(false));
      await dispatch(ProjectValidationActions.validateProject(importProjectPath));
      const manifest = getProjectManifest(getState());
      const updatedImportPath = getProjectSaveLocation(getState());
      ProjectDetailsHelpers.fixBibleDataFolderName(manifest, initialBibleDataFolderName, updatedImportPath);

      if (!TargetLanguageHelpers.targetBibleExists(updatedImportPath, manifest)) {
        dispatch(AlertModalActions.openAlertDialog(translate('projects.loading_ellipsis'), true));
        console.log('onlineImport() - generate target bible');
        TargetLanguageHelpers.generateTargetBibleFromTstudioProjectPath(updatedImportPath, manifest);
        dispatch(ProjectInformationCheckActions.setSkipProjectNameCheckInProjectInformationCheckReducer(true));
        await delay(200);
        dispatch(AlertModalActions.closeAlertDialog());
        console.log('onlineImport() - validate project');
        await dispatch(ProjectValidationActions.validateProject(updatedImportPath));
      }

      const renamingResults = {};
      await dispatch(ProjectDetailsActions.updateProjectNameIfNecessary(renamingResults));
      const { projectDetailsReducer: { projectSaveLocation } } = getState();

      if (renamingResults.repoRenamed) {
        dispatch({ type: consts.UPDATE_SOURCE_PROJECT_PATH, sourceProjectPath: projectSaveLocation });
        dispatch({ type: consts.UPDATE_SELECTED_PROJECT_FILENAME, selectedProjectFilename: renamingResults.newRepoName });
        await delay(200);
      }
      await dispatch(ProjectImportFilesystemActions.move());

      if (renamingResults.repoRenamed) {
        await dispatch(ProjectDetailsActions.doRenamePrompting());
      }
      dispatch(MyProjectsActions.getMyProjects());

      // TODO: refactor this onlineImport method to remove project opening logic so we are not duplicating logic.

      const finalProjectPath = getProjectSaveLocation(getState());
      console.log('onlineImport() - project import complete: ' + finalProjectPath);
      await dispatch(openProject(path.basename(finalProjectPath), true));
      dispatch(AlertModalActions.closeAlertDialog());
      resolve();
    } catch (error) { // Catch all errors in nested functions above
      console.log('onlineImport() - import error:', error);
      const errorMessage = FileConversionHelpers.getSafeErrorMessage(error, translate('projects.online_import_error', { project_url: link, toPath: importProjectPath }));
      dispatch(recoverFailedOnlineImport(errorMessage));
      reject(errorMessage);
    }
  }));
});

/**
 * Performs recovery actions to cleanup after a failed online import
 * @param {string} errorMessage - A localized error message to show the user.
 * @returns {Function}
 */
export const recoverFailedOnlineImport = (errorMessage) => (dispatch) => {
  // TRICKY: clear last project first to avoid triggering autos-saving.
  dispatch(closeProject());

  if (errorMessage === tc_MIN_VERSION_ERROR) {
    dispatch(showInvalidVersionError());
  } else {
    dispatch(AlertModalActions.openAlertDialog(errorMessage));
  }
  dispatch(ProjectImportStepperActions.cancelProjectValidationStepper());
  dispatch({ type: 'LOADED_ONLINE_FAILED' });
  dispatch(deleteImportProjectForLink());
};

/**
 * TODO: this does not need to be an action.
 * @description - delete project (for link) from import folder
 */
export function deleteImportProjectForLink() {
  return ((dispatch, getState) => {
    const link = getState().importOnlineReducer.importLink;

    if (link) {
      const gitUrl = Repo.sanitizeRemoteUrl(link);
      let project = Repo.parseRemoteUrl(gitUrl);

      if (project) {
        deleteProjectFromImportsFolder(project.name);
      }
    }
  });
}

export function clearLink() {
  return {
    type: consts.IMPORT_LINK,
    importLink: '',
  };
}

export function getLink(importLink) {
  return {
    type: consts.IMPORT_LINK,
    importLink,
  };
}
