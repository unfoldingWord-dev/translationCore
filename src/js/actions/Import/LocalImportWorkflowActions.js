import React from 'react';
import path from 'path-extra';
import ospath from 'ospath';
import fs from 'fs-extra';
import {ipcRenderer} from 'electron';
import consts from '../ActionTypes';
// actions
import * as BodyUIActions from '../BodyUIActions';
import * as AlertModalActions from '../AlertModalActions';
import * as ProjectMigrationActions from '../Import/ProjectMigrationActions';
import * as ProjectValidationActions from '../Import/ProjectValidationActions';
import * as ProjectImportFilesystemActions from './ProjectImportFilesystemActions';
import * as ProjectImportStepperActions from '../ProjectImportStepperActions';
import * as MyProjectsActions from '../MyProjects/MyProjectsActions';
import * as ProjectLoadingActions from '../MyProjects/ProjectLoadingActions';
// helpers
import * as TargetLanguageHelpers from '../../helpers/TargetLanguageHelpers';
import * as FileConversionHelpers from '../../helpers/FileConversionHelpers';
import {getTranslate, getProjectManifest, getProjectSaveLocation} from '../../selectors';
import * as ProjectMergeActions from '../ProjectMergeActions';

// constants
export const ALERT_MESSAGE = (
  <div>
    No file was selected. Please click on the
    <span style={{color: 'var(--accent-color-dark)', fontWeight: "bold"}}>
      &nbsp;Import Local Project&nbsp;
    </span>
    button again and select the project you want to load.
  </div>
);
const IMPORTS_PATH = path.join(ospath.home(), 'translationCore', 'imports');
const PROJECTS_PATH = path.join(ospath.home(), 'translationCore', 'projects');

/**
 * @description Action that dispatches other actions to wrap up local importing
 */
export const localImport = () => {
  return async (dispatch, getState) => {
    const translate = getTranslate(getState());
    // selectedProjectFilename and sourceProjectPath are populated by selectProjectMoveToImports()
    const {
      selectedProjectFilename,
      sourceProjectPath
    } = getState().localImportReducer;
    let importProjectPath = path.join(IMPORTS_PATH, selectedProjectFilename);
    try {
      // convert file to tC acceptable project format
      await FileConversionHelpers.convert(sourceProjectPath, selectedProjectFilename);
      ProjectMigrationActions.migrate(importProjectPath);
      await dispatch(ProjectValidationActions.validate(importProjectPath));
      const manifest = getProjectManifest(getState());
      importProjectPath = getProjectSaveLocation(getState());
      if (!TargetLanguageHelpers.targetBibleExists(importProjectPath, manifest)) {
        TargetLanguageHelpers.generateTargetBibleFromTstudioProjectPath(importProjectPath, manifest);
        await delay(400);
        await dispatch(ProjectValidationActions.validate(importProjectPath));
      }
      const projectName = getState().localImportReducer.selectedProjectFilename;
      const projectPath = path.join(PROJECTS_PATH, projectName);
      if (fs.pathExistsSync(projectPath)) {
        // The project already exists so we handle merging the import data with existing project data,
        // then delete the project dir and then continue the import that copies the import dir to the projects dir
        dispatch(ProjectMergeActions.handleProjectMerge(projectPath, importProjectPath))
        .then(() => {
          dispatch(continueImport());
        })
        .catch(error => {
          dispatch(cancelImport(error));
        });
      } else {
        dispatch(continueImport());
      }
    } catch (error) { // Catch all errors in nested functions above
      const errorMessage = FileConversionHelpers.getSafeErrorMessage(error, translate('projects.import_error', {fromPath: sourceProjectPath, toPath: importProjectPath}));
      dispatch(cancelImport(errorMessage));
    }
  };
};

/**
 * @description selects a project from the filesystem and moves it to tC imports folder.
 * @param startLocalImport - optional parameter to specify new startLocalImport function (useful for testing).
 * Default is localImport()
 */
export const selectLocalProject = (startLocalImport = localImport) => {
  return (dispatch, getState) => {
    return new Promise(async (resolve) => {
      const translate = getTranslate(getState());
      dispatch(BodyUIActions.dimScreen(true));
      dispatch(BodyUIActions.closeProjectsFAB());
      // TODO: the filter name and dialog text should not be set here.
      // we should instead send generic data and load the text in the react component with localization
      // or at least we could insert the locale keys here.
      await delay(500);
      const options = {
        properties: ['openFile'],
        filters: [
          {name: translate('supported_file_types'), extensions: ['usfm', 'sfm', 'txt', 'tstudio', 'tcore']}
        ]
      };
      let filePaths = ipcRenderer.sendSync('load-local', {options: options});
      dispatch(BodyUIActions.dimScreen(false));
      // if import was cancel then show alert indicating that it was cancel
      if (filePaths && filePaths[0]) {
        dispatch(AlertModalActions.openAlertDialog(translate('projects.importing_local_alert'), true));
        const sourceProjectPath = filePaths[0];
        const selectedProjectFilename = path.parse(sourceProjectPath).base.split('.')[0] || '';
        await delay(100);
        dispatch({type: consts.UPDATE_SOURCE_PROJECT_PATH, sourceProjectPath});
        dispatch({type: consts.UPDATE_SELECTED_PROJECT_FILENAME, selectedProjectFilename});
        await dispatch(startLocalImport());
        resolve();
      } else {
        dispatch(AlertModalActions.closeAlertDialog());
        resolve();
      }
    });
  };
};

const continueImport = () => {
  return dispatch => {
    dispatch(ProjectImportFilesystemActions.move())
    .then(() => {
      dispatch(MyProjectsActions.getMyProjects());
      dispatch(ProjectLoadingActions.displayTools());
    })
    .catch(error => {
      dispatch(cancelImport(error));
    });
  };
};

const cancelImport = errorMessage => {
  return dispatch => {
    if (errorMessage) {
      dispatch(AlertModalActions.openAlertDialog(errorMessage));
    }
    // clear last project must be called before any other action.
    // to avoid triggering auto-saving.
    dispatch(ProjectLoadingActions.clearLastProject());
    dispatch(ProjectImportStepperActions.cancelProjectValidationStepper());
    // remove failed project import
    dispatch(ProjectImportFilesystemActions.deleteProjectFromImportsFolder());
  };
};

const delay = (ms) => {
  return new Promise((resolve) =>
    setTimeout(resolve, ms)
  );
};
