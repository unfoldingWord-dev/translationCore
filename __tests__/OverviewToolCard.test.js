/* eslint-env jest */
jest.unmock('fs-extra');
jest.unmock('adm-zip');
import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ToolCard from '../src/js/components/home/overview/ToolCard';
import renderer from 'react-test-renderer';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import * as LocalImportWorkflowActions from '../src/js/actions/Import/LocalImportWorkflowActions';
const project = '__tests__/fixtures/project/tstudio_project/abu_tit_text_reg.tstudio';
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('../src/js/components/home/toolsManagement/ToolCardProgress', () => 'ToolCardProgress');

describe('Tool Card component Tests', () => {
  let store;
  beforeEach(() => {
    // create a new store instance for each test
    store = mockStore({
      localImportReducer: {
        selectedProjectFilename: "abu_tit_text_reg",
        sourceProjectPath: project
      },
      settingsReducer: {
        developerMode: ''
      },
      projectDetailsReducer: {
        manifest: {
          project: {
            id: 'tit'
          },
          target_language: {
            id: 'abu'
          },
          'resource': {
            id: 'reg'
          }
        }
      }
    });
  });

  it('ToolCard Before Project Loaded', () => {
    const props = {
      reducers: {
        toolsReducer: {
          currentToolTitle: '',
          currentToolName: ''
        },
        projectDetailsReducer: {
          currentProjectToolsProgress: '',
          projectSaveLocation: ''
        }
      },
      actions: {}
    };
    const component = renderer.create(
      <MuiThemeProvider>
        <ToolCard {...props} />
      </MuiThemeProvider>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    store.dispatch(LocalImportWorkflowActions.localImport());

    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('ToolCard After Project Loaded', () => {
    const title = 'Current Tool Title';
    const name = 'ToolName';
    const progress = 0.5;
    const props = {
      reducers: {
        toolsReducer: {
          currentToolTitle: title,
          currentToolName: name
        },
        projectDetailsReducer: {
          currentProjectToolsProgress: {
            [name]: progress
          }
        }
      },
      actions: {}
    };
    const component = renderer.create(
      <MuiThemeProvider>
        <ToolCard {...props} />
      </MuiThemeProvider>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});