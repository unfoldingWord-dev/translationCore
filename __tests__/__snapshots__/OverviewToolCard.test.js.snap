// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tool Card component Tests ToolCard After Project Loaded 1`] = `
<div
  style={
    Object {
      "flex": 1,
    }
  }
>
  <span>
    Current Tool 
    <a
      onClick={[Function]}
    >
      Change Tool
    </a>
  </span>
  <div
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0,0,0,0)",
        "background": "var(--background-color-light)",
        "backgroundColor": "#ffffff",
        "borderRadius": 2,
        "boxShadow": "0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12)",
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "fontFamily": "Roboto, sans-serif",
        "marginTop": "5px",
        "muiPrepared": true,
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "zIndex": 1,
      }
    }
  >
    <div
      style={
        Object {
          "paddingBottom": 8,
        }
      }
    >
      <div
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 14,
            "muiPrepared": true,
            "padding": 16,
          }
        }
      >
        <div
          style={
            Object {
              "display": "flex",
              "justifyContent": "space-between",
              "margin": "-10px 0 -24px 0",
            }
          }
        >
          <div
            style={
              Object {
                "color": "lightgray",
                "height": "110px",
                "margin": "-6px 20px 0 -16px",
                "overflow": "hidden",
                "width": "100px",
              }
            }
          >
            <span
              className="glyphicon glyphicon-check"
              style={
                Object {
                  "fontSize": "120px",
                  "margin": "-10px 0 0 -25px",
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "width": "400px",
              }
            }
          >
            <strong
              style={
                Object {
                  "fontSize": "x-large",
                }
              }
            >
              Current Tool Title
            </strong>
            <ToolCardProgress
              progress={0.5}
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tool Card component Tests ToolCard Before Project Loaded 1`] = `
<div
  style={
    Object {
      "flex": 1,
    }
  }
>
  <span>
    Current Tool 
    <a />
  </span>
  <div
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0,0,0,0)",
        "background": "var(--background-color-light)",
        "backgroundColor": "#ffffff",
        "borderRadius": 2,
        "boxShadow": "0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12)",
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "fontFamily": "Roboto, sans-serif",
        "marginTop": "5px",
        "muiPrepared": true,
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "zIndex": 1,
      }
    }
  >
    <div
      style={
        Object {
          "paddingBottom": 8,
        }
      }
    >
      <div
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 14,
            "muiPrepared": true,
            "padding": 16,
          }
        }
      >
        <div
          style={
            Object {
              "paddingBottom": "0px",
              "textAlign": "center",
            }
          }
        >
          Select a tool
          <br />
          <button
            className="btn-prime"
            disabled={true}
            onClick={[Function]}
            style={
              Object {
                "marginBottom": "0",
              }
            }
          >
            Tool
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tool Card component Tests ToolCard Before Project Loaded 2`] = `
<div
  style={
    Object {
      "flex": 1,
    }
  }
>
  <span>
    Current Tool 
    <a />
  </span>
  <div
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0,0,0,0)",
        "background": "var(--background-color-light)",
        "backgroundColor": "#ffffff",
        "borderRadius": 2,
        "boxShadow": "0 1px 6px rgba(0, 0, 0, 0.12),
         0 1px 4px rgba(0, 0, 0, 0.12)",
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "fontFamily": "Roboto, sans-serif",
        "marginTop": "5px",
        "muiPrepared": true,
        "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "zIndex": 1,
      }
    }
  >
    <div
      style={
        Object {
          "paddingBottom": 8,
        }
      }
    >
      <div
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 14,
            "muiPrepared": true,
            "padding": 16,
          }
        }
      >
        <div
          style={
            Object {
              "paddingBottom": "0px",
              "textAlign": "center",
            }
          }
        >
          Select a tool
          <br />
          <button
            className="btn-prime"
            disabled={true}
            onClick={[Function]}
            style={
              Object {
                "marginBottom": "0",
              }
            }
          >
            Tool
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
